MINI_2016/Homeworks/Fifth/h5_mvb.Rmd

```diff
@@ -0,0 +1,196 @@
+            FinelineCount = n_distinct(FinelineNumber)
+            )
+# prop.table(table(tmp$Day))
+# prop.table(table(data$Day))
+head(data)
+
+dataDepartments = import %>%
+  group_by(TripType, VisitNumber, DepartmentDescription) %>%
+  summarise(count = n()) %>%
+  spread(DepartmentDescription, count, fill=0)
+head(dataDepartments)
+
+prodsPerDept = import %>%
+  group_by(VisitNumber, DepartmentDescription) %>%
+  summarise(
+    BoughtPerDepartment = sum(ScanCount >=0),
+    ReturnedPerDepartment = sum(ScanCount < 0)
+  )
+head(prodsPerDept)
+
+
+merged.data <- merge(data, dataDepartments, by=c("TripType", "VisitNumber"))
+merged.data <- merge(merged.data, prodsPerDept, by=c("VisitNumber"))
+merged.data <- subset(merged.data, select = -c(DepartmentDescription))
+```
+
+
+#Data Preparation
+Since the dataset is very complex now (with 80 variables and over 300k observations) I need to make some data preparations.
+
+Firstly I decrease the number of observations to 100k and then I will remove the `nearZeroVariance` variables.
+When the dataset is ready I split it into a training and testing set.
+
+```{r, echo=TRUE, message=FALSE, warning=FALSE}
+indxSample <- sample(seq_len(nrow(merged.data)), size = 100000)
+sample <- merged.data[indxSample, ]
+
+prop.table(table(sample$TripType))
+
+nzv <- nearZeroVar(sample)
+names(sample)[nzv]
+data <- sample[, -nzv]
+
+data$TripType <- factor(data$TripType)
+levels(data$TripType)
+
+names(data) = make.names(names(data), unique = FALSE)
+
+set.seed(1313)
+indxTrain <- createDataPartition(y = data$TripType, p = 0.75)
+train <- data[indxTrain$Resample1,]
+test <- data[-indxTrain$Resample1,]
+```
+
+
+#Classification
+##Random Forest
+```{r, echo=TRUE, message=FALSE, warning=FALSE, fig.align='center'}
+rf <- randomForest(TripType ~ ., data=train)
+scores <- predict(rf, test, type = "prob")
+
+myScores <- sapply(1:nrow(test), function(i){
+  scores[i, test$TripType[i]]
+})
+
+perf <- mean(-log(pmax(myScores,0.05)))
+perf
+
+scoresClass <- predict(rf, test, type = "class")
+rfConfusionMatrix <- confusionMatrix(scoresClass, test$TripType)
+
+
+varImpPlot(rf)
+importance(rf)
+```
+
+
+##Random Forest top values by importance
+```{r, echo=TRUE, message=FALSE, warning=FALSE}
+sortedImportance=order(-rf$importance[,1])
+tops=rownames(rf$importance)[sortedImportance][1:15][1:15]
+
+tops <- append(tops, "TripType")
+tops
+subTrain <- train[tops]
+subTest <- test[tops]
+
+rfTop <- randomForest(TripType ~ ., data=subTrain)
+scoresTop <- predict(rfTop, subTest, type = "prob")
+
+myScoresTop <- sapply(1:nrow(subTest), function(i){
+  scoresTop[i, subTest$TripType[i]]
+})
+
+perfTop <- mean(-log(pmax(myScoresTop,0.05)))
+perfTop
+
+scoresTopClass <- predict(rfTop, subTest, type = "class")
+rfTopConfusionMatrix <- confusionMatrix(scoresTopClass, subTest$TripType)
+```
+
+
+#Results
+```{r, echo=TRUE, message=FALSE, warning=FALSE}
+#Random Forest
+rfConfusionMatrix$overall
```

Comment:
It is better to show the whole confusion matrix: all predictions in a table and the metrics. Then we can view trends more easily.